test/python/compiler/test_transpiler.py

```diff
@@ -928,6 +928,23 @@ def test_no_infinite_loop(self, optimization_level):
 
         self.assertEqual(out, expected)
 
+    @data(0, 1, 2, 3)
+    def test_transpile_preserves_circuit_metadata(self, optimization_level):
+        """Verify that transpile preserves circuit metadata in the output."""
+        circuit = QuantumCircuit(2, metadata=dict(experiment_id='1234', execution_number=4))
+        circuit.h(0)
+        circuit.cx(0, 1)
+
+        cmap = [[1, 0], [1, 2], [2, 3], [4, 3], [4, 10],
+                [5, 4], [5, 6], [5, 9], [6, 8], [7, 8],
+                [9, 8], [9, 10], [11, 3], [11, 10],
+                [11, 12], [12, 2], [13, 1], [13, 12]]
+
+        res = transpile(circuit, basis_gates=['id', 'p', 'sx', 'cx'],
+                        coupling_map=cmap,
+                        optimization_level=optimization_level)
```

Comment:
Does this also test that `metadata` is properly passed around when converting to and from `dagdependency`?

Reply:
Err, no it it's not since the only pass using dagdependency is the template matching which isn't in the default optimization levels. I'll add a dedicated converter test verifying that it's preserved through conversion with dagdependency.

Reply:
Test coverage added in: https://github.com/Qiskit/qiskit-terra/pull/5270/commits/06140cd5e76de8372108127c792df8a2819a07a5